INTERNALS.md

```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+    +-----------+   +-----------+   +-----------+
+         |                                ^
+         |                     deliver    |     
+         `--------------------------------´
+
+
+## Job lifecycle
+
+Jobs (in particular, delivery jobs) can be created as _scheduled_ ("run after a
+specified time in the future") or _instant_ ("run as soon as possible").
+
+Workers pick up _instant_ jobs and change their status to _pending_. Each
+worker has at most one pending job.
+
+If a worker dies and the job subsequently recovered ("scrubbed"), the job gets
+demoted from _pending_ back to _instant_ status.
+
+The job executor can also ask for a job to be retried in the future (by raising
+a specific exception); in this case the job is demoted back to _scheduled_
+status.
+
+If a job succeeds, the job and all references are simply removed; there is
+no materialisation of the terminal status.
+
+               promote         acquire           ack
+    +-----------+   +-----------+   +-----------+   +-----------+
+    |           |   |           |   |           |   |           |
+    | Scheduled |-->|   Ready   |-->|  Pending  |-->|  Deleted  |
+    |           |   |           |   |           |   |           |
+    +-----------+   +-----------+   +-----------+   +-----------+
+         ^                                |
+         |                                | nack
+         \--------------------------------/
 
-### Data layout
 
-All Redis keys are namespaced, under `rm:` by default.
+Note that we enforce job uniqueness: only one instance of a job may be enqueued
+at any point, irrespective of state. The control layer make it a no-op to
+enqueue a duplicate job when a copy is already scheduled, ready, or pending.
+
+
+## Data layout
+
+Any mention of "UID" refers to a 20-character, Base64-encoded string
+(representing a 120-bit number), intending to be globally unique.
+
+All timestamps are represented as integers, milliseconds since the Unix epoch.
+
 
 `topics`
 
   The set of all topic names.
 
 `subscribers`
 
-  The set of all subscriber UUIDs.
+  The set of all subscriber tokens.
 
-`topics:{uuid}`
+`topics:{token}`
 
-  The set of topic names subscribed to by subscriber `uuid`.
+  The set of topic names subscribed to by subscriber `{token}`.
 
-`subscribers:{topic}`
+`subscribers:{name}`
 
-  The set of subscriber UUIDs having subscribed to topic `name`.
+  The set of subscriber tokens having subscribed to topic `{name}`.
 
 `topic:{name}`
 
   A hash containing metadata has about a topic. Keys:
   - `publisher`: the UUID of the (singly authorized) publisher
   - `counter`: the cumulative number of events received
 
-`subscriber:{uuid}`
+`subscriber:{token}`
 
   A hash of subscription medatata. Keys:
   - `callback`: the URL to send events to.
-  - `timeout`: how long to defer event delivery for batching purposes.
+  - `timeout`: how long to defer event delivery for batching purposes (aka deadline).
```

Comment:
I realise this isn't a change, but I'm not sure `timeout` is a very good name here as it would lead me to believe it's related to the receive timeout rather than the event staleness. Maybe `max_staleness` or something along those lines would be a more descriptive key?